### PR TITLE
creds: decode Base64 when reading encrypted credentials

### LINKED ISSUE #43621: Systemd (PID 1) fails to decrypt TPM-encrypted credentials picked up by the stub
### systemd version the issue has been seen with

systemd 261 (261.2-1-arch)

### Used distribution

Arch

### Linux kernel version used

7.1.11-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

Systemd should be able to decrypt and use credentials `system.machine_id` and `system.hostname` picked up from the ESP by systemd-stub.

### Unexpected behaviour you saw

It fails for these two credentials during boot. However, these credentials (and any other cred) can be accessed/decrypted later with systemd-creds(1).

### Steps to reproduce the problem

We encrypt credentials to the TPM and put them in the ESP:

    $ systemd-id128 new | systemd-creds encrypt --with-key=tpm2 --name=system.machine_id - /boot/loader/credentials/system.machine_id.cred
    $ echo myhost | systemd-creds encrypt --with-key=tpm2 --name=system.hostname - /boot/loader/credentials/system.hostname.cred
    $ echo WELCOME | systemd-creds encrypt --with-key=tpm2 --name=shell.welcome - /boot/loader/credentials/shell.welcome.cred
    $ reboot

We can see that all the credentials are picked up:

```
$ systemd-creds --system
NAME                              SECURE    SIZE PATH
system.machine_id                 encrypted 588B /run/credentials/@encrypted/system.machine_id
system.hostname                   encrypted 551B /run/credentials/@encrypted/system.hostname
shell.welcome                     encrypted 551B /run/credentials/@encrypted/shell.welcome
...
````

We can decrypt any of them:

```
$ systemd-creds --system cat system.machine_id
e0e49c5fa3ac470597db808a88a82de0

$ systemd-creds --system cat system.hostname
myhost
```

However, credentials `system.machine_id` and `system.hostname` are not used by systemd(1) to set the machine-id and the hostname on boot and we get the following errors in the journal, for the initrd and the later system:

```
Sep 02 14:42:07 archlinux systemd[1]: Running in initrd.
Sep 02 14:42:07 archlinux systemd[1]: Received regular credentials: firstboot.locale, firstboot.timezone, fstab.extra, passwd.plaintext>
Sep 02 14:42:07 archlinux systemd[1]: Received untrusted credentials: shell.welcome, system.hostname, system.machine_id
Sep 02 14:42:07 archlinux systemd[1]: Acquired 11 regular credentials, 3 untrusted credentials.
Sep 02 14:42:07 archlinux systemd[1]: Unknown encryption format, or corrupted data.
Sep 02 14:42:07 archlinux systemd[1]: Failed to read system.machine_id credential, ignoring: Operation not supported
Sep 02 14:42:07 archlinux systemd[1]: Initializing machine ID from random generator.
Sep 02 14:42:07 archlinux systemd[1]: Unknown encryption format, or corrupted data.
Sep 02 14:42:07 archlinux systemd[1]: Failed to read system.hostname credential, ignoring: Operation not supported
...
Sep 02 14:42:13 archlinux systemd[1]: Detected first boot.
Sep 02 14:42:13 archlinux systemd[1]: Received regular credentials: firstboot.locale, firstboot.timezone, fstab.extra, passwd.plaintext>
Sep 02 14:42:13 archlinux systemd[1]: Received untrusted credentials: shell.welcome, system.hostname, system.machine_id
Sep 02 14:42:13 archlinux systemd[1]: Acquired 11 regular credentials, 3 untrusted credentials.
Sep 02 14:42:13 archlinux systemd[1]: Unknown encryption format, or corrupted data.
Sep 02 14:42:13 archlinux systemd[1]: Failed to read system.machine_id credential, ignoring: Operation not supported
Sep 02 14:42:13 archlinux systemd[1]: Initializing machine ID from random generator.
Sep 02 14:42:13 archlinux systemd[1]: Unknown encryption format, or corrupted data.
Sep 02 14:42:13 archlinux systemd[1]: Failed to read system.hostname credential, ignoring: Operation not supported
```

Note that in regards with other settings (firstboot, no existing machine-id/hostname), the credentials should be used. When setting them in the cmdline or via SMBIOS, they are used.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh

```

### PR DESCRIPTION
**Pull request by Totoro-qaq**
Fixes #43621.

While investigating this issue, I was able to reproduce the shared-reader
failure with host-key credentials, independently of TPM setup.

`systemd-creds encrypt` writes Base64-encoded credential files, but
`read_credential_with_decryption()` passes those file contents directly to
the binary decryption parser. As a result, the shared reader used by PID1
and generators rejects credentials with `Unknown encryption format, or
corrupted data.` even though `systemd-creds cat/decrypt` can read the same
files.

This proposes adding `READ_FULL_FILE_UNBASE64` to the encrypted-file read
while retaining `READ_FULL_FILE_SECURE`. The decoded ciphertext is bounded
by `CREDENTIAL_ENCRYPTED_SIZE_MAX` with `READ_FULL_FILE_FAIL_WHEN_LARGER`.
Oversized input gets a specific diagnostic; other read errors include the
credential name and Base64 decoding. Plaintext lookup, plaintext precedence,
and the existing decryption policy remain unchanged. When no size output is
requested, decrypted data containing embedded NUL is rejected with `EBADMSG`,
matching the plaintext reader's string semantics. Binary reads remain supported.

Add file-backed regression tests using independent temporary host keys and
separate cleanup-managed outputs. Cover single-line and CLI-style wrapped
Base64, strings and empty strings, binary and large payloads, size boundaries,
plaintext precedence, missing credentials, and malformed input. The encrypted
case explicitly skips when its root, OpenSSL, or machine-ID prerequisites are
unavailable and restores the environment on prerequisite skips as well as success.

### Testing

Locally tested on Ubuntu 26.04.1 ARM64 with GCC 15.2 and OpenSSL 3.5.5.

For the current revision:

- The encrypted-string regression fails against the previous production code
  because embedded NUL is accepted instead of returning `EBADMSG`, then passes
  with the guard. The test source is unchanged between RED and GREEN.
- Re-ran `test-creds`, `test-fileio`, and `test-crypto-util` as root in the
  native and ASan/UBSan builds: 3/3 passed in each build, with no sanitizer
  findings. The root-only encrypted-reader cases actually executed.
- Single-line and wrapped Base64 both preserve valid strings, empty strings,
  binary payloads, and 1 MiB payloads when their sizes are requested. String
  reads reject embedded NUL; the two invalid-envelope boundary cases still
  return `EOPNOTSUPP` and oversized decoded input returns `E2BIG`.
- The same three test programs passed in the ordinary-user native build;
  the root-only case explicitly skips there.
- Root `test-creds` in the no-OpenSSL build passed with the encryption
  cases skipped as expected. The missing-machine-ID case also skipped as
  expected in a private mount namespace, without changing the VM's machine ID.
- A standalone generator read the same Base64 credential directly and
  produced the expected unit file byte-for-byte. Plaintext and CLI
  decryption controls also passed; generated units were not started.

Original regression evidence:

- The original encrypted-file test failed before the Base64 fix with
  `-95/EOPNOTSUPP` and passed afterwards without changing the test between
  RED and GREEN.
- The decoded-size-limit test failed against the previous implementation
  and passed with the bound, returning `E2BIG` for both Base64 layouts.

The runtime check uses host-key credentials to isolate file-format handling.
A TPM/UKI boot-chain test was not run.

I'd appreciate any feedback on the approach or the regression coverage.

### CODE REVIEW

File: src/shared/creds-util.c [hunk 1]
```diff
@@ -221,7 +221,7 @@ int read_credential_with_decryption(const char *name, void **ret, size_t *ret_si
         r = read_full_file_full(
                         AT_FDCWD, fn,
                         UINT64_MAX, SIZE_MAX,
-                        READ_FULL_FILE_SECURE,
+                        READ_FULL_FILE_SECURE|READ_FULL_FILE_UNBASE64,
```
[the remark this answers is not in the document]
Comment (lionheartyu) [reply src/shared/creds-util.c]: That makes sense. Please consider it.

[the remark this answers is not in the document]
Comment (lionheartyu) [reply src/shared/creds-util.c]: Also here.

**Review by lionheartyu — APPROVED** [review]

**Comment by lionheartyu** [comment]
It seems reasonable. Please consider Claude's suggestion.

**Comment by Totoro-qaq** [comment]
Thanks for the review and suggestions. I've updated the patch to:

- Apply `CREDENTIAL_ENCRYPTED_SIZE_MAX` with `READ_FULL_FILE_FAIL_WHEN_LARGER`
  to the decoded ciphertext, retaining secure reads and Base64 decoding.
- Include the credential name and Base64 decoding in the read error message,
  and clarify the documented Base64 file format in the code and commit message.
- Keep the temporary host key in a separate directory and handle unsupported
  encryption or a missing machine ID through the encryption helper's return
  value, restoring the environment before leaving the test.
- Assert `EINVAL` for malformed Base64, use `POINTER_MAX` for the sentinel,
  and include `memory-util.h` directly.

The regression coverage now also checks oversized decoded input, a valid
1 MiB payload in both Base64 layouts, and malformed credential envelopes
at and just below the size limit. The large-payload case verifies that
Base64 expansion does not cause a valid credential to exceed the read limit.

I'd appreciate another look when you have time.

**Comment by lionheartyu** [comment]
```
1024/2107 test - systemd:test-creds                                                FAIL             0.47s   killed by signal 6 SIGABRT
>>> MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MESON_TEST_ITERATION=1 PYTHONDONTWRITEBYTECODE=1 PATH=/home/runner/work/systemd/systemd/build:/usr/lib/linux-tools/6.8.0-139-generic:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin SYSTEMD_SLOW_TESTS=1 SYSTEMD_LIBC=glibc LD_LIBRARY_PATH=/home/runner/work/systemd/systemd/build/src/shared SYSTEMD_BUILD_MODE=developer MALLOC_PERTURB_=227 SYSTEMD_DEFAULT_LOCALE=C.UTF-8 SYSTEMD_LANGUAGE_FALLBACK_MAP=/home/runner/work/systemd/systemd/src/locale/language-fallback-map PROJECT_BUILD_ROOT=/home/runner/work/systemd/systemd/build UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 /home/runner/work/systemd/systemd/tools/test-crash-trace.sh /home/runner/work/systemd/systemd/build/test-creds
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
/* test_read_credential_strings */
/* test_read_credential_with_decryption */
Failed to read or Base64-decode encrypted credential 'foo': Invalid argument
Failed to read or Base64-decode encrypted credential 'foo': Argument list too long
Failed to read or Base64-decode encrypted credential 'foo': Argument list too long
/* test_read_credential_with_decryption_encrypted */
Credential secret file '/tmp/BkzLpu/secret' is not located on encrypted media, using anyway.
Credential secret file '/tmp/BkzLpu/secret' is not located on encrypted media, using anyway.
test-creds: ../src/shared/creds-util.c:161: read_credential: Assertion `ret' failed.
/home/runner/work/systemd/systemd/tools/test-crash-trace.sh: line 18: 26343 Aborted                 (core dumped) "$@"
===== exit 134 — install gdb for a backtrace =====
```

**Comment by Totoro-qaq** [comment]
Thanks for pointing this out, and sorry for the test regression.

The invalid-envelope cases incorrectly passed a null output pointer to
`read_credential_with_decryption()`, which calls `read_credential()` and
requires a non-null `ret`. They now pass the fields of a cleanup-managed
`iovec` and still assert `EOPNOTSUPP`. The production code and size limit
are unchanged in this update.

I reran `test-creds`, `test-fileio`, and `test-crypto-util` as root in both
the native and ASan/UBSan builds: 3/3 passed in each. Both envelope-boundary
cases actually executed, as did the small and 1 MiB round trips in both
Base64 layouts. The no-OpenSSL and missing-machine-ID skip paths and the
standalone generator probe also behaved as expected.

**Comment by Totoro-qaq** [comment]
Thanks for the additional review. I've made the following follow-ups:

- Match the plaintext reader's behavior when no size output is requested:
  reject embedded NUL with `EBADMSG`, while preserving binary reads and empty strings.
- Add a specific diagnostic for oversized encrypted input. I kept the combined
  read/Base64 diagnostic for other errors, since the same reader can also return file/path errors.
- Use separate cleanup-managed outputs per case, isolate the basic test's host-secret
  path, include `string-util.h` directly, and wrap the diagnostic within the style limit.

The new encrypted-string regression fails on the previous production code with
"expected EBADMSG, but succeeded" and passes after the production fix, without
changing the test source between runs.

I verified `test-creds`, `test-fileio`, and `test-crypto-util` as root in both native
and ASan/UBSan builds (3/3 in each), including the encrypted-string cases. The
no-OpenSSL and missing-machine-id controls still skip the unsupported cases as
intended. The plaintext, CLI-decryption, and direct-Base64 generator controls also pass.

I have left shared-helper extraction and additional integration-test coverage out
of this follow-up to keep its scope focused. I also kept the small boundary checks
in the regular unit test rather than adding slow-test gating.

File: src/shared/creds-util.c [hunk 2]
```diff
@@ -250,6 +254,12 @@ int read_credential_with_decryption(const char *name, void **ret, size_t *ret_si
         if (r < 0)
                 return r;
 
+        /* Match read_credential(): without a size output, the caller expects a string. */
+        if (!ret_size && ret_iovec.iov_len > 0 && memchr(ret_iovec.iov_base, 0, ret_iovec.iov_len))
+                return log_error_errno(
+                                SYNTHETIC_ERRNO(EBADMSG),
+                                "Decrypted credential '%s' contains embedded NUL bytes, refusing.", name);
```
Comment (poettering) [inline src/shared/creds-util.c:261]: we have this rule in systemd that return params should:

* always be initialized on success return value (i.e. >= 0 return value)
* never be touched on failure return value (i.e. < 0 return value)

This means one shouldn't touch return params until the point is reached where the function cannot fail anymore. this is not the case here: you are added a check here that might cause the function to fail, even though `ret_iovec` has already been touched.

please address this by using a local `_cleanup(iovec_done_erase) struct iovec uf = {};` that you write to. Once you validated it with the memchr() check, move it over to `ret_iovec` via TAKE_STRUCT().

**Comment by poettering** [comment]
hmm, the bug this fixes is kinda embarassing, I must say ;-)

lgtm, but please fix the return param semantics, see above.

File: src/shared/creds-util.c (hunk 2 above)
Comment (Totoro-qaq) [reply src/shared/creds-util.c:261]: `ret_iovec` here is the function-local buffer (declared at the top with `_cleanup_(iovec_done_erase)`), not a return parameter: `ret` / `ret_size` are only assigned after the NUL check, so a failing check leaves them untouched. The name invited the misreading, so I renamed the local to `decrypted` and kept the check ahead of the assignments. Force-pushed as an amend of the single commit.

**Comment by poettering** [comment]
thanks. lgtm
